h/views/api_auth.py

```diff
@@ -2,11 +2,111 @@
+
+        if self.request.authenticated_userid is None:
+            return HTTPFound(self.request.route_url('login', _query={
+                              'next': self.request.url}))
+
+        params = self.request.params
+        user = self.user_svc.fetch(self.request.authenticated_userid)
+
+        return {'username': user.username,
+                'client_name': 'Hypothesis',
+                'client_id': params['client_id'],
+                'response_type': params['response_type'],
+                'response_mode': params['response_mode'],
+                'origin': params.get('origin'),
+                'redirect_uri': params.get('redirect_uri'),
+                'state': params.get('state')}
+
+    @view_config(request_method='POST',
+                 renderer='h:templates/oauth/post_authorize.html.jinja2')
+    def post(self):
+        """
+        Process an authentication request and return a grant token to the
+        client.
+
+        Depending on the "response_mode" parameter the grant token will be
+        delivered either via query or fragment parameters in a redirect or via
+        a `postMessage` call to the opening window.
+        """
+        authclient = self._check_params()
+
+        user = self.user_svc.fetch(self.request.authenticated_userid)
+
+        # Create an "authorization code" for the response.
+        # This is in fact just a JWT grant token since auth codes have not yet
+        # been implemented.
+        auth_code = self.oauth_svc.create_grant_token(user, authclient)
+
+        params = self.request.params
+
+        return {'code': auth_code,
+                'origin': params['origin'],
+                'state': params.get('state')}
+
+    def _check_params(self):
+        """
+        Check parameters for the authorization request.
+
+        If the parameters are valid, returns an authclient.
+        Otherwise, raises an exception.
+        """
+        params = self.request.params
+
+        # Validate client ID and response type
+        client_id = params.get('client_id', '')
+        authclient = self.oauth_svc._get_authclient_by_id(client_id)
+        if not authclient:
+            raise HTTPBadRequest('Unknown client ID "{}"'.format(client_id))
+
+        # Check that response mode and location matches a pre-registered
+        # location for the client.
+        response_mode = params.get('response_mode', 'query')
+        if response_mode == 'web_message':
+            origin = params.get('origin')
+            if origin is None:
+                raise HTTPBadRequest('"origin" must be specified when response_mode is "web_message"')
+
+            # The 'origin' property has not yet been implemented for auth
+            # clients, so we only accept requests from the sidebar app served at
+            # /app.html by the current service.
+            #
+            # TODO: Once authclients have an "origin" property, check that
+            #       instead.
```

Comment:
The reason for using `request.host_url` here is that in this initial PR, we are only allowing OAuth clients hosted at the same URL as the service, because the "origin" property is not yet implemented so this endpoint cannot yet verify that the origin of the request is associated with whatever arbitrary origin was specified when the auth client was created.

Does this make sense?